File: Platform/mo-1-common/src/lang-java/org/mo/com/net/FSocketOutput.java
package org.mo.com.net;

import java.io.OutputStream;
import org.mo.com.lang.FFatalError;
import org.mo.com.lang.FObject;

//============================================================
// <T>网络输出流。</T>
//============================================================
public class FSocketOutput
      extends FObject
{
   public static int BLOCK_SIZE = 8196;

   // 网络链接
   protected FSocket _socket;

   // 输出流
   protected OutputStream _outputStream;

   // 写入长度
   protected int _writeLength;

   //============================================================
   // <T>构造网络输出流。</T>
   //
   // @param socket 网络链接
   //============================================================
   public FSocketOutput(FSocket socket){
      try{
         _socket = socket;
         _outputStream = socket.nativeOutputStream();
      }catch(Exception e){
         throw new FFatalError(e);
      }
   }

   //============================================================
   // <T>获得网络链接。</T>
   //
   // @return 网络链接
   //============================================================
   public FSocket socket(){
      return _socket;
   }

   //============================================================
   // <T>写入数据内容。</T>
   //
   // @param data 数据
   //============================================================
   public void write(byte[] data){
      try{
         if(data != null && data.length > 0){
            _writeLength += data.length;
            _outputStream.write(data);
         }
      }catch(Exception e){
         throw new FFatalError(e);
      }
   }

   //============================================================
   // <T>写入数据内容。</T>
   //
   // @param data 数据
   // @param length 长度
   //============================================================
   public void write(byte[] data,
                     int length){
      try{
         if(data != null && data.length > 0){
            _writeLength += data.length;
            _outputStream.write(data, 0, length);
         }
      }catch(Exception e){
         throw new FFatalError(e);
      }
   }

   //============================================================
   // <T>写入数据内容。</T>
   //
   // @param data 数据
   // @param offset 位置
   // @param length 长度
   //============================================================
   public void write(byte[] data,
                     int offset,
                     int length){
      try{
         if(data != null && data.length > 0){
            _writeLength += length;
            _outputStream.write(data, offset, length);
         }
      }catch(Exception e){
         throw new FFatalError(e);
      }
   }

   //============================================================
   // <T>刷新数据流。</T>
   //============================================================
   public void flush(){
      try{
         _outputStream.flush();
      }catch(Exception e){
         throw new FFatalError(e);
      }
   }

   //============================================================
   // <T>关闭输入流。</T>
   //============================================================
   public void close(){
      try{
         _outputStream.close();
      }catch(Exception e){
         throw new FFatalError(e);
      }
   }
   //public boolean write(int nValue, int nDigits) {
   //return write(FStringUtil.leftPad(Integer.toString(nValue), nDigits, "0"));
   //}
   //
   //   public void write(int value){
   //      try{
   //         _writeLength += 4;
   //         _outputStream.write(value);
   //      }catch(Exception e){
   //         throw new FFatalError(e);
   //      }
   //   }
   //
   //   public boolean write(String value){
   //      try{
   //         if(value != null){
   //            byte[] bytes = value.getBytes();
   //            if(bytes.length > 0){
   //               _writeLength += bytes.length;
   //               _outputStream.write(bytes);
   //               return true;
   //            }
   //         }
   //         return false;
   //      }catch(Exception e){
   //         throw new FFatalError(e);
   //      }
   //   }
   //
   //   public boolean write(StringBuffer value){
   //      try{
   //         if(value != null){
   //            byte[] bytes = value.toString().getBytes();
   //            if(bytes.length > 0){
   //               _writeLength += bytes.length;
   //               _outputStream.write(bytes);
   //               return true;
   //            }
   //         }
   //         return false;
   //      }catch(Exception e){
   //         throw new FFatalError(e);
   //      }
   //   }
   //
   //   //   public boolean write(FInputStream oInputStream, int nLength) {
   //   //      try {
   //   //         if (oInputStream != null && nLength > 0) {
   //   //            int nReadLength = 0;
   //   //            int nTranslateLenth = 0;
   //   //            byte[] arCache = new byte[BLOCK_SIZE];
   //   //            while (true) {
   //   //               nReadLength = oInputStream.read(arCache);
   //   //               nTranslateLenth += nReadLength;
   //   //               if (nReadLength > 0) {
   //   //                  _outputStream.write(arCache, 0, nReadLength);
   //   //                  if (nTranslateLenth >= nLength) {
   //   //                     break;
   //   //                  }
   //   //               } else if (nReadLength == -1) {
   //   //                  break;
   //   //               }
   //   //            }
   //   //            _outputStream.flush();
   //   //            _writeLength += nLength;
   //   //            return true;
   //   //         }
   //   //         return false;
   //   //      } catch (Exception e) {
   //   //         throw new FFatalException(e);
   //   //      }
   //   //   }
   //
   //   public int writeLength(){
   //      return _writeLength;
   //   }
}

File: Utiliy/Common/MuCommon/Net/Socket/FSocketOutput.cs
using System;
using System.Net.Sockets;
using System.Text;
using MO.Common.IO;

namespace MO.Common.Net.Sockets
{
   //============================================================
   // <T>Socket输出数据流。</T>
   //============================================================
   public class FSocketOutput
   {
      private FSocket _socket;

      //============================================================
      // <T>构造输出数据流。</T>
      //============================================================
      public FSocketOutput() {
      }

      //============================================================
      // <T>构造输出数据流并绑定Socket。</T>
      //
      // @param socket 指定的Socket
      //============================================================
      public FSocketOutput(FSocket socket) {
         _socket = socket;
      }

      //============================================================
      public FSocket Socket {
         get { return _socket; }
         set { _socket = value; }
      }

      //============================================================
      // <T>将一个整数写入输出流中。</T>
      //
      // @param value 写入的数据
      // @return 已发送到Socket的字节数
      //============================================================
      public int Write(int value) {
         byte[] bytes = BitConverter.GetBytes(value);
         return _socket.NativeSocket.Send(bytes, SocketFlags.None);
      }

      //============================================================
      // <T>将一个字符串写入输出流中。</T>
      //
      // @param value 写入的数据
      // @return 已发送到Socket的字节数
      //============================================================
      public int Write(string value) {
         return Write(value, Encoding.ASCII);
      }

      //============================================================
      // <T>将一个字符串按照指定字符编码格式写入输出流中。</T>
      //
      // @param value 写入的数据
      // @param encoding 字符编码
      // @return 已发送到Socket的字节数
      //============================================================
      public int Write(string value, Encoding encoding) {
         byte[] bytes = encoding.GetBytes(value);
         return _socket.NativeSocket.Send(bytes, SocketFlags.None);
      }

      //============================================================
      // <T>将一个字符串按照指定字符编码格式写入输出流中，
      // 并且在末尾加换行符。</T>
      //
      // @param value 写入的数据
      // @param encoding 字符编码
      // @return 已发送到Socket的字节数
      //============================================================
      public int WriteLine(string line, Encoding encoding) {
         byte[] bytes = encoding.GetBytes(line + RFile.LINE);
         return _socket.NativeSocket.Send(bytes, SocketFlags.None);
      }

      //============================================================
      // <T>将一个byte数组写入输出流中。</T>
      //
      // @param bytes 写入的数据
      // @return 已发送到Socket的字节数
      //============================================================
      public int Write(byte[] bytes) {
         return _socket.NativeSocket.Send(bytes, SocketFlags.None);
      }

      //============================================================
      // <T>将一个byte数组的指定部分写入输出流中。</T>
      //
      // @param bytes 写入的数据
      // @param offset 指定偏移量
      // @param count 要写入的字节数
      // @return 已发送到Socket的字节数
      //============================================================
      public int Write(byte[] bytes, int offset, int count) {
         return _socket.NativeSocket.Send(bytes, offset, count, SocketFlags.None);
      }
   }
}
